packages/world/src/systems/BuildSystem.sol

```diff
@@ -147,52 +168,53 @@ contract BuildSystem is System {
       SeedGrowth._setFullyGrownAt(base, uint128(block.timestamp) + ctx.buildType.getTimeToGrow());
     } else if (ctx.buildType.hasExtraDrops()) {
       DisabledExtraDrops._set(base, true);
+    } else if (ctx.buildType == ObjectTypes.ForceField) {
+      (EntityId existing,) = ForceFieldUtils.getForceField(ctx.coord);
+      require(!existing._exists(), "Force field overlaps with another force field");
+      ForceFieldUtils.setupForceField(base, ctx.coord);
     }
   }
 
   /**
    * @dev Validates builds against force fields and calls build hooks for programs
    */
-  function _requireBuildsAllowed(BuildContext memory ctx, EntityId base, Vec3[] memory coords, bytes calldata extraData)
-    internal
-  {
+  function _requireBuildsAllowed(BuildContext memory ctx, EntityId base, Vec3[] memory coords) internal {
     for (uint256 i = 0; i < coords.length; i++) {
-      if (ctx.buildType == ObjectTypes.ForceField) {
-        (EntityId existing,) = ForceFieldUtils.getForceField(coords[i]);
-        require(!existing._exists(), "Force field overlaps with another force field");
-        ForceFieldUtils.setupForceField(base, coords[i]);
-        // Return early as it is not possible for a new forcefield to have a program
-        return;
-      }
-
-      (ProgramId program, EntityId target, EnergyData memory energyData) = ForceFieldUtils.getHookTarget(coords[i]);
-
-      if (!program.exists()) {
-        continue;
-      }
-
-      program.hook({ caller: ctx.caller, target: target, revertOnFailure: energyData.energy > 0, extraData: extraData })
-        .onBuild({
+      Vec3 coord = coords[i];
+
+      _hook(ctx, coord).onBuild({
         entity: base,
-        coord: coords[i],
+        coord: coord,
         slotType: ctx.slotType,
         objectType: ctx.buildType,
         orientation: ctx.orientation
       });
     }
   }
+
+  /// @dev Returns the pre-built hook to avoid stack too deep issues
+  function _hook(BuildContext memory ctx, Vec3 coord) internal returns (Hook memory) {
+    (ProgramId program, EntityId target, EnergyData memory energyData) = ForceFieldUtils.getHookTarget(coord);
+    return program.hook({
+      caller: ctx.caller,
+      target: target,
+      revertOnFailure: energyData.energy > 0,
+      extraData: ctx.extraData
+    });
+  }
 }
 
 library BuildLib {
   function _executeBuild(BuildContext memory ctx) public returns (EntityId, Vec3[] memory) {
-    uint128 energyLeft = transferEnergyToPool(ctx.caller, Math.min(ctx.callerEnergy, BUILD_ENERGY_COST));
-    if (energyLeft == 0) {
+    if (transferEnergyToPool(ctx.caller, ctx.energyCost) == 0) {
```

Comment:
Don't we still need to do Math.min? what if player has less energy than the energyCost?

Reply:
we do that when building the context